### PR TITLE
heartbeat support

### PR DESCRIPTION
**Pull request by trenouf**

### CODE REVIEW

File: lib/src/main/java/io/ably/lib/transport/ConnectionManager.java [hunk 1]
```diff
@@ -109,6 +109,8 @@ public StateIndication(ConnectionState state, ErrorInfo reason, boolean fallback
 		put(ConnectionState.failed, new StateInfo(ConnectionState.failed, false, false, true, false, 0, REASON_FAILED));
 	}};
 
+	public long maxIdleInterval;
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/transport/ConnectionManager.java]: public?

File: lib/src/main/java/io/ably/lib/transport/ITransport.java [hunk 2]
```diff
@@ -39,6 +39,11 @@
 		String connectionKey;
 		String connectionSerial;
 		Mode mode;
+		boolean heartbeats;
+
+		public TransportParams() {
+			heartbeats = true; // default to requiring Ably heartbeats
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/transport/ITransport.java]: We're using `/* .... */ ` for all comments please.

File: lib/src/main/java/io/ably/lib/transport/ConnectionManager.java (hunk 1 above)
Comment (trenouf) [inline lib/src/main/java/io/ably/lib/transport/ConnectionManager.java]: Done

File: lib/src/main/java/io/ably/lib/transport/ITransport.java (hunk 2 above)
Comment (trenouf) [inline lib/src/main/java/io/ably/lib/transport/ITransport.java]: Done

**Comment by mattheworiordan** [comment]
Thanks, looks good, would have been nice to have seen a test to ensure we don't introduce a regression down the line which could be pretty nasty given none of our tests would probably detect it now, and in production connections would just be terminated.

**Comment by mattheworiordan** [comment]
@trenouf @paddybyers I think we should add a test to cover this. Happy to add an issue that we look at later though.  Please advise

**Comment by trenouf** [comment]
So @mattheworiordan are you thinking a test that does nothing for 2 mins and checks that the connection stays?

**Comment by mattheworiordan** [comment]
> So @mattheworiordan are you thinking a test that does nothing for 2 mins and checks that the connection stays?

Nope, I would expect that we could change the heartbeat period by perhaps creating a simple WS server that responds with a Connected message that indicates heartbeats are expected say every 1s. You could then a) not send them to see if it fails when expected, b) send them at that rate for say 15 seconds and ensure we've not disconnected the connection.  It will require a bit of jiggery, but where necessary we do this in all libraries as a regression in the future of this kind would be quite devastating.

**Comment by mattheworiordan** [comment]
@paddybyers @trenouf I've just realised we've merged this and tagged a release, yet as far as I am aware, we don't yet support native WebSocket heartbeats on realtime i.e. this is not yet released as it's part of 0.9.  Does that not mean this change would result in Java clients thinking that the connection has gone away, or does this change ensure that both native heartbeats and Ably protocol heartbeats will work as a means to detect liveness?

**Comment by paddybyers** [comment]
> as far as I am aware, we don't yet support native WebSocket heartbeats on realtime

Realtime has been sending ws ping frames for a long time.

The 0.9 feature is to request that these are the only pings, ie protocol heartbeats are suppressed.

**Comment by mattheworiordan** [comment]
> Realtime has been sending ws ping frames for a long time.

Ok, good, thanks

**Pull request merged**